### DISCUSSION TITLE
Parameterized multi-row inserts with queryRaw (createMany)

### DISCUSSION THREAD

**Discussion by ghost** [prisma/prisma#4177] [Q&A] [answered]
Since Prisma does not currently have a createMany function, I was trying to use postgresql's support for multi-row inserts with queryRaw to insert multiple records with one INSERT statement. However, I have not been able to find a way to do parameterized multi-row inserts with prisma. I noticed that pg-format has a way to do basically what I am looking for,

```ts
const pg = require('pg');
const format = require('pg-format');

let users = [['test@example.com', 'Fred'], ['test2@example.com', 'Lynda']];
let query1 = format('INSERT INTO users (email, name) VALUES %L returning id', users);
```
which gets translated into this statement:
`INSERT INTO users (email, name) VALUES ('test@example.com', 'Fred'), ('test2@example.com', 'Lynda') returning id`

Is there a way to do this with Prisma? Have I missed something or do I need to use something like pg-format? A createMany function that basically does the same thing doesn't seem like too much to ask for as a feature.

**Comment by ryands17** [comment, marked as answer]
Hey @jejacobsendev 👋 

Considering a schema like this:

```prisma
model User {
  id Int @id @default(autoincrement())
  name String
  email String
}
```

You can do a batch insert like this:

```ts
let values = users.map((user) => `('${user.name}', '${user.email}')`)
let query = `insert into "User" (name, email) values ${values.join(',')}`
await prisma.executeRaw(query)
```

Also do follow [this](https://github.com/prisma/prisma/issues/3619) request for more updates.

**Reply by ghost** [reply]
Thanks! One of the values is my model is an array of strings (text[]) so I had to add extra brackets to get it to work with an array and a join with "," separator to get it to work in the case of empty string values
```
 => `('{"${sides.join('", "')}"}', ${vocabId})`
```
The only other thing is that for some reason, the way the array values show up in pgAdmin is a bit weird. It shows it like: 
{hello,"english greeting","something else"} where the first value in the array doesn't have quotes but every other value does. Another array  of the same type in my db, none of the values have quotes.  Not sure if this is something I need to be worried about or if it could cause issues at any point. But other than that every thing works fine.

**Reply by ryands17** [reply]
For an array, you could do this:

```ts
  const users = [
    { name: 'user 1', strings: ['text 1', 'text 2'] },
    { name: 'user 2', strings: ['text 3', 'text 4'] },
  ]

  let values = users.map(
    user => `('${user.name}', '{${user.strings.map(u => `"${u}"`).join(',')}}')`
  )
  let query = `insert into "User" ("name", "strings") values ${values.join(
    ','
)}`

await prisma.$executeRaw(query)
```

**Reply by kasir-barati** [reply]
I guess since 2020 they have changed how `queryRaw` and `executeRaw` works a lot since this code:
```ts
const values = ['val1', 'val2'].map((title) => {
  return `(gen_random_uuid(), NOW(), NOW(), '${title}')`;
});
const feeds = await prismaClient.$queryRaw`INSERT INTO public.feeds(id, created_at, updated_at, title)
VALUES ${values.join(', ')}
RETURNING *;`;
```

Throws this tantrum:
```shell
Something went wrong! PrismaClientKnownRequestError: 
Invalid `prisma.$executeRaw()` invocation:


Raw query failed. Code: `42601`. Message: `ERROR: syntax error at or near "$1"`
    at Mn.handleRequestError (/home/kasir/projects/react/node_modules/.pnpm/@prisma+client@5.21.1_prisma@5.21.1/node_modules/@prisma/client/runtime/library.js:121:7753)
    at Mn.handleAndLogRequestError (/home/kasir/projects/react/node_modules/.pnpm/@prisma+client@5.21.1_prisma@5.21.1/node_modules/@prisma/client/runtime/library.js:121:7061)
    at Mn.request (/home/kasir/projects/react/node_modules/.pnpm/@prisma+client@5.21.1_prisma@5.21.1/node_modules/@prisma/client/runtime/library.js:121:6745)
    at async l (/home/kasir/projects/react/node_modules/.pnpm/@prisma+client@5.21.1_prisma@5.21.1/node_modules/@prisma/client/runtime/library.js:130:9633)
    at Repo.bulkFeeds (/home/kasir/projects/react/apps/api/src/utils/repo.util.mts:25:19)
    at <anonymous> (/home/kasir/projects/react/apps/api/prisma/seed.ts:10:3) {
  code: 'P2010',
  clientVersion: '5.21.1',
  meta: { code: '42601', message: 'ERROR: syntax error at or near "$1"' }
```

And when I change it to use `Prisma.sql` and `Prisma.join` like this:
```ts
const values = ['val1', 'val2'].map((title) => {
  return Prisma.sql`(gen_random_uuid(), NOW(), NOW(), '${title}')`;
});
const feeds = await this.prismaClient.$queryRaw`INSERT INTO public.feeds(id, created_at, updated_at, title)
VALUES ${Prisma.join(values)}
RETURNING *;`;

console.log(feeds[0]);
```

I get the following log:

```js
{
  id: '4073aa85-07eb-4447-b099-1218a9e5fc63',
  created_at: 2024-11-01T22:35:59.290Z,
  updated_at: 2024-11-01T22:35:59.290Z,
  title: '$1'
}
```

IDK how on earth it decided to put "$1" inside title as its value :exploding_head:.

**Comment by Be-P** [comment]
Hi guys, I'm trying to use this approach. The reason is that I want to add "ON CONFLICT DO NOTHING" to the insert.
The lines of code that are trying  to achieve a "raw-create-many" follow:
```
const values = list.map(
    (tx) =>
      `(uuid_generate_v4(), '${tx.fiatCoin}', '${tx.cryptoCoin}', '${tx.ForeingKeyEntity.id}')`,
  );
const query = Prisma.sql`insert into public."GatewayAllowedTransaction" (id, "fiatCoin", "cryptoCoin", "foreignKeyEntityId" ) values ${values.join(',')}`;
await this.prisma.$executeRaw(query);
```
When running the above code I get: `db error: ERROR: syntax error at or near "$1"`, but the interesting fact is that just copying the log of the executed query and params to a bare-bone psql shell, and executing it, works! 

Edit1. I've tryied to substitute `uuid_generate_v4()` with a application-level uuid but it doesn't solve the issue

Solution:  you've to use `executeRawUnsafe` and do validation avoiding SQLInjection yourself if you want to use the solution in the accepted  answer. the `executeRaw` do not accepts raw strings anymore, and if you switch to Prisma.sql templating it will not work as expected

**Reply by carlossantillana** [reply]
I ran into the same issue; however there is a way around it without opening yourself up to SQL injections by using executeRawUnsafe.
Given the example by [ryands17](https://github.com/prisma/prisma/discussions/4177#discussioncomment-124611)
you can update the code to 

```
const values = users.map((user) => Prisma.sql`('${user.name}', '${user.email}')`)
const query = Prisma.sql`insert into "User" (name, email) values ${Prisma.join(values)}`
await prisma.executeRaw(query)
```
Which leverages [SQL Template Tag](https://github.com/blakeembrey/sql-template-tag) to nest tagged templates to keep both Prisma happy and your code base free of sql injections

**Reply by rohanrajpal** [reply]
@carlossantillana Thanks a ton!

**Reply by kasir-barati** [reply]
For some reason it is not working anymore: 
```ts
const values = ['val1', 'val2'].map((title) => {
  return Prisma.sql`(gen_random_uuid(), NOW(), NOW(), '${title}')`;
});
const query = Prisma.sql`INSERT INTO public.feeds(id, created_at, updated_at, title)
VALUES ${Prisma.join(values)}
RETURNING *;`;
const feeds = await this.prismaClient.$queryRaw<Feed[]>(query);

console.log(feeds[0]);
```

logs:

```js
{
  id: '233a83e6-7132-4532-bcce-40879d64296c',
  created_at: 2024-11-01T22:42:17.288Z,
  updated_at: 2024-11-01T22:42:17.288Z,
  title: '$1'
}
```

**Comment by kasir-barati** [comment]
For me @carlossantillana's suggestion and @ryands17's answer did not work as everyone of would have expected it to as I explained it already. Thus I had to use `queryRawUnsafe`:

```ts
const values = ['val1', 'val2'].map((title) => {
  return `(gen_random_uuid(), NOW(), NOW(), '${title}')`;
});
const feeds = await this.prismaClient.$queryRawUnsafe<
  Feed[]
>(`INSERT INTO public.feeds(id, created_at, updated_at, title)
VALUES ${values.join(', ')}
RETURNING *;`);

console.log(feeds[0]);
```

This logs
```js
{
  id: 'ac7bfc40-1d0c-4992-8c0e-ff39916aa365',
  created_at: 2024-11-01T22:59:21.994Z,
  updated_at: 2024-11-01T22:59:21.994Z,
  title: 'val1'
}
```

If you know what is wrong [here](https://github.com/prisma/prisma/discussions/4177#discussioncomment-11126519) please enlighten me too.
